network/dns.go

```diff
@@ -29,43 +30,56 @@ const (
 	openDNSRequestLimit = 3 * time.Second
 )
 
-func getDNSRequestCacheKey(pid int, fqdn string) string {
-	return strconv.Itoa(pid) + "/" + fqdn
+func getDNSRequestCacheKey(pid int, fqdn string, qType uint16) string {
+	return strconv.Itoa(pid) + "/" + fqdn + dns.Type(qType).String()
 }
 
 func removeOpenDNSRequest(pid int, fqdn string) {
 	openDNSRequestsLock.Lock()
 	defer openDNSRequestsLock.Unlock()
 
-	key := getDNSRequestCacheKey(pid, fqdn)
-	_, ok := openDNSRequests[key]
-	if ok {
-		delete(openDNSRequests, key)
-		return
-	}
+	// Delete PID-specific requests.
+	delete(openDNSRequests, getDNSRequestCacheKey(pid, fqdn, dns.TypeA))
+	delete(openDNSRequests, getDNSRequestCacheKey(pid, fqdn, dns.TypeAAAA))
 
+	// If process is known, also check for non-attributed requests.
 	if pid != process.UnidentifiedProcessID {
-		// check if there is an open dns request from an unidentified process
-		delete(openDNSRequests, unidentifiedProcessScopePrefix+fqdn)
+		delete(openDNSRequests, getDNSRequestCacheKey(process.UnidentifiedProcessID, fqdn, dns.TypeA))
+		delete(openDNSRequests, getDNSRequestCacheKey(process.UnidentifiedProcessID, fqdn, dns.TypeAAAA))
 	}
 }
 
 // SaveOpenDNSRequest saves a dns request connection that was allowed to proceed.
-func SaveOpenDNSRequest(conn *Connection) {
+func SaveOpenDNSRequest(q *resolver.Query, rrCache *resolver.RRCache, conn *Connection) {
+	// Only save requests that actually went out to reduce clutter.
+	if rrCache.ServedFromCache {
+		return
+	}
+
+	// Try to "merge" A and AAAA requests into the resulting connection.
+	// Save others immediately.
+	switch uint16(q.QType) {
+	case dns.TypeA, dns.TypeAAAA:
+	default:
+		conn.Save()
+		return
+	}
+
```

Comment:
We should save the query type and other query metadata (maybe the whole query) along the connection. Maybe as a nilable `DNSContext` or similar. Otherwise we won't be able to display any meaningful information about DNS requests except the requested domain ...